File: Data Structures & Algorithms/coin-change/submission-1.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        BIGNUMBA = 10000000
        dp = [-5] * (amount + 1)
        def opt(a):
            if a == 0:
                return 0
            if dp[a] == -5:
                minC = BIGNUMBA
                for c in coins:
                    if c > a or opt(a - c) == -1:
                        continue
                    minC = min(minC, opt(a - c) + 1)
                if minC == BIGNUMBA:
                    dp[a] = -1
                else:
                    dp[a] = minC
            return dp[a]
        return opt(amount)

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.cpp
class Solution {
public:
    int findMin(vector<int> &nums) {
        auto l{0};
        auto r{static_cast<int>(nums.size() - 1)};
        while (l < r) {
            auto m{l + (r - l) / 2};
            if (nums[l] < nums[r]) {
                return nums[l];
            }
            if (nums[l] < nums[m]) {
                l = m + 1;
            } else if (nums[m] < nums[r]) {
                r = m;
            } else {
                return std::min(nums[l], nums[r]);
            }
        }
        return nums[l];
    }
};
